### ISSUE TITLE
Class signature verification error for multi-release jars

### ISSUE THREAD

**Issue by Yeregorix** [MinecraftForge/SecureModules#3]
I discovered that a multi-release signed jar loaded in a `SecureModuleClassLoader` triggers a verification error.

`java.lang.SecurityException: class "org.spongepowered.configurate.objectmapping.RecordFieldDiscoverer"'s signer information does not match signer information of other classes in the same package`

I made a MRE : https://github.com/Yeregorix/ForgeBrokenSigner
Import the project (Java 16+) and run Gradle tasks `runNormal` and `runWithForgeBootstrap` to compare the results with and without `SecureModuleClassLoader`.

In this example the jar is `configurate-core` version `4.1.2` which is a multi-release with signed classes. 
If you open the jar you will find two versions of class `RecordFieldDiscoverer`.
~~One for Java 8 an one for Java 16. The manifest only contains the signature of the first class, but at runtime the second class is loaded.~~
This causes an error when using `SecureModuleClassLoader` but doesn't in the standard class loader.

**Comment by LexManos** [comment]
Interesting, part of the point of SM's classloading system is that the normal java classloader post JPMS does not verify or store signatures. So it's no surprise that it doesn't error on normal java. 

I'll have to take a look at this when I get back into town. (Next week ish) Do you know the mechanics of how this jar is getting signed? Most importantly why it does not have the Multi Release files signed? Last I looked signtool just signed everything in the jar including versioned classes.

Worth testing if SM works fine with a complete set of signatures. Otherwise I will have to do some thinking on how best to detect this case.

**Comment by Yeregorix** [comment]
Looking at Configurate's build logic, it seems to be signed via Gradle, which delegates to Ant, which delegates to JDK's jarsigner CLI tool.

See:
https://github.com/SpongePowered/Configurate/blob/24e08cf3aac2ec28c9f9d3dc6eb4a123f72efabf/build-logic/src/main/groovy/org/spongepowered/configurate/build/SignJarTask.groovy#L39
https://ant.apache.org/manual/Tasks/signjar.html

Just to test, I got rid of previous signatures and re-applied manually `jarsigner` on the jar. The manifest looks the same and the error persists. ~~Not signing version specific classes seems to be the standard behavior.~~

I took a look at the JAR file specification and I see no mention of multi-release jars with signatures. I don't even know what would be the name of the entry for the signature of the Java 16 RecordFieldDiscoverer.
I guess it would be `META-INF/versions/16/org/spongepowered/configurate/objectmapping/RecordFieldDiscoverer.class`.

~~Also, even if we managed to generate a complete set of signatures for Configurate, the issue would persist for any other multi-release jar where we don't have control over the build chain.~~

**Comment by LexManos** [comment]
Yes I understand that it would cause issues for anyone who doesn't sign their versioned files. I'm mearly trying to understand if it's a simple mistake on your end that is causing it to not sign the versioned files, or if its something explicitly filtered by signtool. 

And yes that is the path I would expect to see the signature in as the manifest is simply a list of hashes for each file in the archive. Versioned files just being another file in the archive. 

And ya, I doubt the spec would say anything about the signatures of multi release jars as that feature was introduced in J9 with the JPMS which got rid of runtime jar signature verification. Hence part of the reason SM exists.

Looking at your scripts I see two cases [here](https://github.com/SpongePowered/Configurate/blob/24e08cf3aac2ec28c9f9d3dc6eb4a123f72efabf/build-logic/src/main/groovy/org.spongepowered.configurate.build.publishing.gradle#L52) which uses the `jar` task. And [here](https://github.com/SpongePowered/Configurate/blob/24e08cf3aac2ec28c9f9d3dc6eb4a123f72efabf/build-logic/src/main/groovy/org.spongepowered.configurate.build.shadow-component.gradle#L39) Which updates? it to use the `shadowJar` task. Which one is actually used, and are you sure it has the versioned files in it? Not meant in an accusatory way just in a "I'm on vacation in the middle of utah on my phone so can't verify myaelf" kinda way.

Anyways, is this file that is erroring the unsigned versioned file or another signed class in the same package? I already capture the first signature data of the first class loaded in each package and propagate it. It needs some more work apparently, and needs better logging but you can see it [here](https://github.com/MinecraftForge/SecureModules/blob/master/src/main/java/net/minecraftforge/securemodules/SecureModuleClassLoader.java#L625) I think the issues is I didnt test unsigned being loaded before signed.

**Comment by Yeregorix** [comment]
So... turns out I gave incorrect information. Sorry! I thought the signatures of the versioned classes were missing but they are all present in the jar and the `jarsigner` CLI seems to work correctly. 

For the explanation, the manifest is ~700 lines so I was searching for keyword `RecordFieldDiscoverer` with a text editor instead of scrolling all lines. There was only a result for the non-versioned class that's why I thought the other was missing. Today I tried searching for keyword `META-INF` and found this:

```
Name: META-INF/versions/16/org/spongepowered/configurate/objectmapping/R
 ecordFieldDiscoverer.class
SHA-256-Digest: mrEvsPD9eqtD8LbHLUGtFd9JuBL+8cSytqSMKJDPPuE=
```
The line break is why I didn't find this entry previously. 😅 

I feel a bit dumb but at least now we know that all signatures are present.

**Comment by LexManos** [comment]
Ha, it happens. Okay so if that is the case. Then I just need to add some special casing to support versioned files. Should be simple enough when I get time. Like I said I'm on vacation right now so wont be for a bit. Unless I find somewhere with both electricity and wifi :p 

But good to know that I wasnt going crazy with how jarsigner worked.

**Comment by LexManos** [comment]
Yup it was basically what I expected. Tho it did have me go down the road of figuring out how normal java verifies signatures. And they do it on reading the raw resource. I think in the future I'll refactor the project to do that as well, which *should* simplify things and bring us more in line with vanilla behavior.

I also saw that the way it's implemented now we verify the transformed bytes so it means that no classes can have coremods apply to them. BUT we silently ignore any invalid signatures which means as long as someone tampers with everything in your package they can modify to their heart content.

Basically, lots of learning the code, and lots of ideas to address in the future. However your issue should be resolved as it now takes name overrides into account when finding the correct hashes.
